### PR TITLE
[flutter_tools] Fix roll dev script, add tests

### LINKED ISSUE #54774: [flutter_tools] roll_dev.dart
The `roll_dev.dart` script is still publishing tags with the `x.y.z-dev.m.n` format, whereas it should be publishing using the new(er) format `x.y.z-m.n.pre`.

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

Though the tool was [updated to support the new release tag format](https://github.com/flutter/flutter/pull/54217) (e.g. `1.2.3-4.5.pre`), the `roll_dev.dart` script was never updated. This PR updates the script to both read and write the new tag format. The next dev release will have to be manually tagged before this script will work, as it depends on being able to read the last tagged release to determine what the next tag should be (if it is run today, it will fail saying it could not find the last release). I will be manually tagging the next release.

I changed this script to exit with usage instructions if an explicit `--commit` was not provided. It previously defaulted to `'upstream/master'`, which seems like an unsafe feature (especially since part of the script is doing a hard reset to said commit). I can't think of a situation where we would want to arbitrarily release whatever is at the tip of master.

Added unit tests, and manually tested with my own repo fork (I locally tagged `1.18.0-1.2.pre`, changed the `kUpstreamRemote` to `christopherfujino/flutter` and ran the script: https://github.com/christopherfujino/flutter/releases).

## Related Issues

Fixes https://github.com/flutter/flutter/issues/54774.

## Tests

I added unit tests to verify the regex and document which tag formats are and are not supported.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

### CODE REVIEW

File: dev/tools/lib/roll_dev.dart [hunk 1]
```diff
@@ -40,7 +41,7 @@ void main(List<String> args) {
     kCommit,
     help: 'Specifies which git commit to roll to the dev branch.',
     valueHelp: 'hash',
-    defaultsTo: 'upstream/master',
+    defaultsTo: null, // This option is required
```
Comment (jonahwilliams) [inline dev/tools/lib/roll_dev.dart:44]: Default value will be null anyway

Comment (jonahwilliams) [reply dev/tools/lib/roll_dev.dart:44]: Also, I would add that the option is required in the help text

**Review by jonahwilliams — COMMENTED** [review]
I don't have a good record of being able to look at regular expression and tell if its correct. I would recommend finding an online regular expression generator (give it a regexp, it generates a conforming string) and create a large set of valid tags to test.

Then, tweak the regular expression so it is subtly wrong and assert that those fail.

Basically a psuedo-fuzz test

File: dev/tools/lib/roll_dev.dart (hunk 1 above)
Comment (christopherfujino) [reply dev/tools/lib/roll_dev.dart:44]: Yeah, I explicitly added a null so people wouldn't add a default later. I will update the help text, though

**Comment by christopherfujino** [comment]
> I don't have a good record of being able to look at regular expression and tell if its correct. I would recommend finding an online regular expression generator (give it a regexp, it generates a conforming string) and create a large set of valid tags to test.
> 
> Then, tweak the regular expression so it is subtly wrong and assert that those fail.
> 
> Basically a psuedo-fuzz test

I added unit tests :)

**Comment by jonahwilliams** [comment]
I would suggest adding more examples of valid/invalid tags to those tests

File: dev/tools/lib/roll_dev.dart (hunk 1 above)
Comment (christopherfujino) [reply dev/tools/lib/roll_dev.dart:44]: done

Comment (jonahwilliams) [reply dev/tools/lib/roll_dev.dart:44]: Makes sense!

**Comment by christopherfujino** [comment]
> I would suggest adding more examples of valid/invalid tags to those tests

added all the premutations i could think of (i'm clearly a bad fuzzer)

**Comment by jonahwilliams** [comment]
I would plug your regex into https://onlinestringtools.com/generate-string-from-regex to see if it thinks of any new ones

**Comment by christopherfujino** [comment]
> I would plug your regex into https://onlinestringtools.com/generate-string-from-regex to see if it thinks of any new ones

Cool!

File: dev/tools/lib/roll_dev.dart [hunk 2]
```diff
@@ -165,23 +167,26 @@ void main(List<String> args) {
...
   return getGitOutput(
-    'describe --match *.*.*-dev.*.* --first-parent --long --tags',
+    'describe --match $glob --first-parent --long --tags',
     'obtain last released version number',
   );
 }
 
 Match parseFullTag(String version) {
-  final RegExp versionPattern = RegExp(r'^([0-9]+)\.([0-9]+)\.([0-9]+)-dev\.([0-9]+)\.([0-9]+)-([0-9]+)-g([a-f0-9]+)$');
+  final RegExp versionPattern = RegExp(
```
Comment (jmagman) [inline dev/tools/lib/roll_dev.dart:178]: Can you add a comment with an example of a tag that should match?

File: dev/tools/lib/roll_dev.dart [hunk 3]
```diff
@@ -165,23 +167,26 @@ void main(List<String> args) {
...
 
 String getVersionFromParts(List<int> parts) {
   assert(parts.length == 5);
   final StringBuffer buf = StringBuffer()
     ..write(parts.take(3).join('.'))
-    ..write('-dev.')
-    ..write(parts.skip(3).join('.'));
+    ..write('-')
+    ..write(parts.skip(3).take(2).join('.'))
+    ..write('.pre');
```
Comment (jmagman) [inline dev/tools/lib/roll_dev.dart:189]: Here too so we don't have to guess what's getting skipped and taken.

File: dev/tools/lib/roll_dev.dart (hunk 2 above)
Comment (christopherfujino) [reply dev/tools/lib/roll_dev.dart:179]: done

File: dev/tools/lib/roll_dev.dart (hunk 3 above)
Comment (christopherfujino) [reply dev/tools/lib/roll_dev.dart]: PTAL

**Review by jmagman — APPROVED** [review]
LGTM

**Review by jonahwilliams — APPROVED** [review]
LGTM _also_

**Pull request merged**